pkg/manager/impl/reservation_manager.go

```diff
@@ -0,0 +1,188 @@
+package impl
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/flyteorg/flytestdlib/promutils"
+	"github.com/flyteorg/flytestdlib/promutils/labeled"
+
+	errors2 "github.com/flyteorg/datacatalog/pkg/errors"
+	"github.com/flyteorg/datacatalog/pkg/repositories"
+	"github.com/flyteorg/datacatalog/pkg/repositories/models"
+	"github.com/flyteorg/datacatalog/pkg/repositories/transformers"
+
+	"github.com/flyteorg/datacatalog/pkg/manager/interfaces"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/datacatalog"
+)
+
+type reservationMetrics struct {
+	scope                        promutils.Scope
+	reservationAcquiredViaCreate labeled.Counter
+	reservationAcquiredViaUpdate labeled.Counter
+	reservationAlreadyInProgress labeled.Counter
+	makeReservationFailure       labeled.Counter
+	getTagFailure                labeled.Counter
+}
+
+type NowFunc func() time.Time
+
+type reservationManager struct {
+	repo               repositories.RepositoryInterface
+	reservationTimeout time.Duration
+	now                NowFunc
+	systemMetrics      reservationMetrics
+}
+
+func NewReservationManager(
+	repo repositories.RepositoryInterface,
+	reservationTimeout time.Duration,
+	nowFunc NowFunc, // Easier to mock time.Time for testing
+	reservationScope promutils.Scope,
+) interfaces.ReservationManager {
+	systemMetrics := reservationMetrics{
+		scope: reservationScope,
+		reservationAcquiredViaCreate: labeled.NewCounter(
+			"reservation_acquired_via_create",
+			"Number of times a reservation was acquired via create",
+			reservationScope),
+		reservationAcquiredViaUpdate: labeled.NewCounter(
+			"reservation_acquired_via_update",
+			"Number of times a reservation was acquired via update",
+			reservationScope),
+		reservationAlreadyInProgress: labeled.NewCounter(
+			"reservation_already_in_progress",
+			"Number of times we try of acquire a reservation but the reservation is in progress",
+			reservationScope,
+		),
+		makeReservationFailure: labeled.NewCounter(
+			"make_reservation_failure",
+			"Number of times we failed to make reservation",
+			reservationScope,
+		),
+		getTagFailure: labeled.NewCounter(
+			"get_tag_failure",
+			"Number of times we failed to get tag",
+			reservationScope,
+		),
+	}
+
+	return &reservationManager{
+		repo:               repo,
+		reservationTimeout: reservationTimeout,
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's reserve a spot to work on
+			// generating the artifact.
+			status, err := r.makeReservation(ctx, request)
```

Comment:
what happens if reservations fails with already exists.

Ideally the `Get` or `Reserve` should be one API and will need a transactional Get.

What you have done is fine too, as long as you have an already exists handling right

Reply:
If the reservation already exists, on https://github.com/flyteorg/datacatalog/blob/c57349c955b680ac8cbc1ce173829e23a62971cd/pkg/manager/impl/reservation_manager.go#L150 we check if the existing reservation has expired and try to grab the reservation via conditional updatee.